addons/mail/static/src/messaging/entity/entity/entity.js

```diff
@@ -0,0 +1,824 @@
+        // Private
+        //--------------------------------------------------------------------------
+
+        /**
+         * @static
+         * @private
+         * @returns {string[]}
+         */
+        static _getListOfClassAttributeNames() {
+            return [];
+        }
+
+        /**
+         * @static
+         * @private
+         * @abstract
+         * @param {Object} data
+         */
+        static _update(data) {}
+
+        /**
+         * Any assignment to class as attribute or stored relation should pass
+         * through this function.
+         *
+         * @static
+         * @private
+         * @param {Object} data
+         */
+        static _write(data) {
+            Object.assign(this.observable, data);
+        }
+
+        /**
+         * @private
+         * @param {Object} data
+         * @returns {string}
+         */
+        _createInstanceLocalId(data) {
+            return _.uniqueId(`${this.constructor.localId}_`);
+        }
+
+        /**
+         * @abstract
+         * @private
+         * @param {Object} data
+         */
+        _update(data) {}
+
+        /**
+         * Any assignment to instance as attribute or stored relation should
+         * pass through this function.
+         *
+         * @private
+         * @param {Object} data
+         */
+        _write(data) {
+            Object.assign(this, data);
+        }
+
+        //--------------------------------------------------------------------------
+        // Internal
+        //--------------------------------------------------------------------------
+
+        /**
+         * @static
+         * @param {mail.messaging.entity.Entity} entity instance or class
+         */
+        static __delete(entity) {
+            const { state } = this.env.store;
+            if (!this.get(entity)) {
+                // Entity has already been deleted.
+                // (e.g. unlinking one of its reverse relation was causal)
+                return;
+            }
+            const relations = entity.relations;
+            const data = {};
+            for (const relationName in relations) {
+                const relation = relations[relationName];
+                if (relation.isCausal) {
+                    switch (relation.type) {
+                        case 'one2one':
+                        case 'many2one':
+                            if (entity[relationName]) {
+                                entity[relationName].delete();
+                            }
+                            break;
+                        case 'one2many':
+                        case 'many2many':
+                            for (const relatedEntity of entity[relationName]) {
+                                relatedEntity.delete();
+                            }
+                            break;
+                    }
+                }
+                data[relationName] = null;
+            }
+            entity.unlink(data);
```

Comment:
There are too many methods that have synonymous names, not even counting the static and/or private versions of them.